### ISSUE TITLE
Trivial memset optimization not applied to loops under -Oz (LoopIdiomRecognize)

### ISSUE THREAD

**Issue by 2ac352c7-76df-4fdf-879e-561e76a0f473** [llvm/llvm-project#50308]
|  |  |
| --- | --- |
| Bugzilla Link | [50964](https://llvm.org/bz50964) |
| Version | 12.0 |
| OS | All |
| CC | @fhahn,@TH3CHARLie |

## Extended Description 
Issue description:
https://llvm.discourse.group/t/loops-not-optimized-under-oz-that-should-have-been/3734/1

**Comment by razetime** [comment]
> I'd be happy to provide help/guidance if anyone is interested in tackling this.
@fhahn Hello, I'm new to the project, and I'd like to solve this issue. I'm currently trying to replicate this bug in C++ so i can figure out what i should modify. This is what I've made to try and reproduce the issue(hope it is correct):

```
#include <vector>
#include <stdint.h>

using namespace std;
int main() {
	vector<uint8_t> v(8);
}
void memzero(vector<uint8_t> vec) {
	for(auto i : vec) {
		i = 0;
	}
}
```


Currently I only know that this has something to do with `llvm-project\llvm\lib\Transforms\Scalar\LoopIdiomRecognize.cpp`. I would like to  get some pointers on where the changes  should be implemented.

**Comment by fhahn** [comment]
I don't think you really need to have a C++ example. It should be sufficient to run `opt  -loop-idiom` on the LLVM IR from https://llvm.godbolt.org/z/YrGqGhnE9 .

As a starting point, you can take a look at what loop-idiom does and why it fails for the example. Adding `-debug` should provide a bit of additional information, but using GDB/additional print statements may also be helpful.

**Comment by razetime** [comment]
Oh, good to know. What I'm currently doing is 
```
opt -loop-idiom -debug OZtest.ll -o OZtestloopid.ll
llc OZtestloopid.ll
clang-14 OZtestloopid.s
```
But I get this error
```
/usr/bin/ld: /lib/x86_64-linux-gnu/crt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
```
I assume I should be running gdb on the result of `clang file.s` and check what is happening there, but i'd like help with rectifying this error first.

**Comment by fhahn** [comment]
There is no need to run `llc` or `clang` after `opt -loop-idiom`.

You can just check the output generated by `opt -loop-idiom -S`. At the moment, there's no `memset` call because the transformation does not trigger. After a fix, `loop-idiom` should create a `memset` and it is sufficient to check the output after `opt` to verify.

**Comment by razetime** [comment]
in [this area](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp#L406) a comment says the following:
```
// The following transforms hoist stores/memsets into the loop pre-header.
// Give up if the loop has instructions that may throw.
```
is this relevant to the problem?

**Comment by razetime** [comment]
Alright, [runOnLoopBlock](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp#L641) is the place where memset patterns are recognized.

**Comment by razetime** [comment]
@fhahn Using the given ir from godbolt and modifying the file here: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/LoopIdiomRecognize.cpp#L628

```
  for (BasicBlock *ExitBlock : ExitBlocks) {
    if (!DT->dominates(BB, ExitBlock)) {
      DEBUG_WITH_TYPE("raze", dbgs() << "appears to not dominate all exit blocks.\n");
      return false;
    }
  }
  ```
here the debug statement executes which i think means that memset optimization application is halted here early. I'd like to know if this is correct and that i'm testing in the right place.

**Comment by razetime** [comment]
@fhahn after some [Phabricator discussion](https://reviews.llvm.org/D116754) and reviewing my existing ideas, I've come up with these ideas for a possible fix:

- Applying loop rotation within `RunOnCountableLoop` if the loop is unrotated
- Changing the pipeline of `-Oz` to add the `loop-rotate` pass before `loop-idiom` based on code analysis

I'd like some pointers on whether these are feasible, and what other approaches i can try to solve this with.

**Comment by fhahn** [comment]
Thanks I responded to the review.

**Comment by razetime** [comment]
@fhahn thanks for the comment. Currently I am writing the following check:
```
  bool isSimpleUnrotated = true;
  SmallVector<BasicBlock *, 8> sm;
  CurLoop->getExitingBlocks(sm);

  auto head =CurLoop->getHeader();
  if(sm.size() == 1 && sm.front() == head && !CurLoop->isRotatedForm()) {
    
    for(Instruction &I: *head) {
      StoreInst *SI = dyn_cast<StoreInst>(&I);
      if (!SI)
        continue;
      switch (isLegalStore(SI))
      {
      case LegalStoreKind::None:
      case LegalStoreKind::Memset:
      case LegalStoreKind::MemsetPattern:
      case LegalStoreKind::Memcpy:
      case LegalStoreKind::UnorderedAtomicMemcpy:
      isSimpleUnrotated = true;
      break;
      default:
        isSimpleUnrotated = false;
        break;
      }
      if (!isSimpleUnrotated) break;
    }
  }
 ```

I'm not sure if i'm assigning false to `IsSimpleUnrotated` at the right places. Also, once this check is done, what would the rotation of the loop look like? will it require extra headers like `LoopRotation.h`?

**Comment by hiraditya** [comment]
Is anyone currently working on this?

**Comment by appujee** [comment]
Working on unrotated loop is probably not the best idea. LoopIdiom may be improved to work on specific cases (like the memset example above) but other optimizations will be skipped. Will it be reasonable to add a flag that allows duplication of loop-header at Oz?

**Comment by hiraditya** [comment]
Working on unrotated loop is probably not the best idea. LoopIdiom may be improved to work on specific cases (like the memset example above) but other optimizations will be skipped. Will it be reasonable to add a flag that allows duplication of loop-header at Oz?

**Comment by hiraditya** [comment]
[Verified](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:16,fontUsePx:'0',j:1,lang:llvm,selection:(endColumn:1,endLineNumber:4,positionColumn:1,positionLineNumber:4,selectionStartColumn:1,selectionStartLineNumber:4,startColumn:1,startLineNumber:4),source:'define+void+@test(i8*+noalias+nonnull+align+1+%25start,+i8*+%25end)+unnamed_addr+%7B%0Aentry:%0A++br+label+%25loop.header%0A%0Aloop.header:%0A++%25ptr.iv+%3D+phi+i8*+%5B+%25start,+%25entry+%5D,+%5B+%25ptr.iv.next,+%25loop.latch+%5D%0A++%25_12.i+%3D+icmp+eq+i8*+%25ptr.iv,+%25end%0A++br+i1+%25_12.i,+label+%25exit,+label+%25loop.latch%0A%0Aloop.latch:%0A++%25ptr.iv.next+%3D+getelementptr+inbounds+i8,+i8*+%25ptr.iv,+i64+1%0A++store+i8+1,+i8*+%25ptr.iv,+align+1%0A++br+label+%25loop.header%0A%0Aexit:%0A++ret+void%0A%7D'),l:'5',n:'0',o:'LLVM+IR+source+%231',t:'0')),k:46.09880919613934,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:opttrunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:llvm,libs:!(),options:'-passes%3D!'default%3COz%3E!'+-enable-loop-header-duplication',overrides:!(),selection:(endColumn:58,endLineNumber:5,positionColumn:58,positionLineNumber:5,selectionStartColumn:58,selectionStartLineNumber:5,startColumn:58,startLineNumber:5),source:1),l:'5',n:'0',o:'+opt+(trunk)+(Editor+%231)',t:'0')),header:(),k:53.90119080386066,l:'4',m:100,n:'0',o:'',s:0,t:'0')),l:'2',m:100,n:'0',o:'',t:'0')),version:4) that it is working.

**Comment by fhahn** [comment]
I think we should only close the issue once the trivial memset gets removed by default when using `-Oz` with Clang/other frontends.

**Comment by ilovepi** [comment]
> I think we should only close the issue once the trivial memset gets removed by default when using `-Oz` with Clang/other frontends.

Sorry. You’re right. I should have just mentioned the issue instead of using ‘Fixes’ in the commit message.

**Comment by hiraditya** [comment]
> Hi [@fhahn](https://github.com/fhahn),
> 
> `opt` no longer recognizes the options (`-loop-rotate -loop-idiom`) specified in the provided link. Could you please share updated pass names to reproduce it?

Can you share the command line you tried?

**Comment by hiraditya** [comment]
from the `opt -help`, 
```
Optimizations available (use "-passes=" for the new pass manager)
```

but I agree this one liner is still not very helpful.

**Comment by hiraditya** [comment]
Ah that's unfortunate.
